### PR TITLE
AWS auth login with multi region STS support

### PR DESCRIPTION
**Pull request by raymonstah**
# Summary
Adds support to the AWS auth plugin so that when the `use_sts_region_from_client` configuration is set, the global STS endpoint (https://sts.amazonaws.com) will be overridden to instead use the [regional STS endpoints](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_temp_enable-regions.html) based on the region provided in the `Authorization` header.

This change allows for DR situations and AWS regional outages by supporting flexibility in where the login region is calling from.

Without this change, clients would run into an error message when logging in if there were a region mismatch between client and server:
```
Error authenticating: Error making API request.

URL: PUT http://127.0.0.1:8200/v1/auth/aws/login
Code: 400. Errors:

* error making upstream request: received error code 403 from STS: <ErrorResponse xmlns="https://sts.amazonaws.com/doc/2011-06-15/">
  <Error>
    <Type>Sender</Type>
    <Code>SignatureDoesNotMatch</Code>
    <Message>Credential should be scoped to a valid region. </Message>
  </Error>
  <RequestId>b749af7f-ed19-4cea-9885-f60d2422880e</RequestId>
</ErrorResponse>
```

# Testing

For the server side, I've configured the AWS auth plugin with the `use_sts_region_from_client` config set to `true`.
```
vault auth enable aws
Success! Enabled aws auth method at: aws/

vault write auth/aws/config/client \
    secret_key="${AWS_SECRET_ACCESS_KEY}" \
    access_key="${AWS_ACCESS_KEY_ID}" \
use_sts_region_from_client=true
Success! Data written to: auth/aws/config/client

vault write auth/aws/role/dev-role-iam \
    auth_type="iam" \
    bound_iam_principal_arn="arn:aws:iam::XXXXXXXXXXX:user/FOOBAR" \
    policies="default"
Success! Data written to: auth/aws/role/dev-role-iam
```

For the client side, I've tested the login command using a bunch of different regions, `auto` region, and no region specified . Responses omitted for brevity.
```
vault login -method=aws  region=us-east-1 role=dev-role-iam
Success! You are now authenticated. The token information displayed below...

vault login -method=aws  region=us-east-2 role=dev-role-iam
Success! You are now authenticated. The token information displayed below...

vault login -method=aws  region=us-west-1 role=dev-role-iam
Success! You are now authenticated. The token information displayed below...

vault login -method=aws  region=us-west-2 role=dev-role-iam
Success! You are now authenticated. The token information displayed below...

vault login -method=aws  region=auto role=dev-role-iam
Success! You are now authenticated. The token information displayed below...

vault login -method=aws role=dev-role-iam
Success! You are now authenticated. The token information displayed below...
```

### CODE REVIEW

File: builtin/credential/aws/path_config_client.go [hunk 1]
```diff
@@ -61,6 +62,12 @@ func (b *backend) pathConfigClient() *framework.Path {
 				Description: "The region ID for the sts_endpoint, if set.",
 			},
 
+			"sts_region_from_client": {
```
Comment (raymonstah) [inline builtin/credential/aws/path_config_client.go:65]: Suggestions appreciated for a better name

Comment (austingebauer) [reply builtin/credential/aws/path_config_client.go:65]: I like `sts_region_from_client`.

File: website/content/api-docs/auth/aws.mdx [hunk 2]
```diff
@@ -65,6 +65,10 @@ capabilities, the credentials are fetched automatically.
 - `sts_region` `(string: "")` - Region to override the default region for making
   AWS STS API calls. Should only be set if `sts_endpoint` is set. If so, should
   be set to the region in which the custom `sts_endpoint` resides.
+- `sts_region_from_client` `(boolean: false)` - If set, overrides both `sts_endpoint` and
```
Comment (austingebauer) [inline website/content/api-docs/auth/aws.mdx:68]: We might mention that this only applies to IAM-based authentication?

File: changelog/21960.txt [hunk 3]
```diff
@@ -0,0 +1,3 @@
+```release-note:improvement
+aws/auth: Adds a new config field `sts_region_from_client` which allows for using dynamic regional sts endpoints based on Authorization header.
```
Comment (austingebauer) [inline changelog/21960.txt:2]: Similar to doc comment, might be good to mention it's for IAM-based auth.

File: builtin/credential/aws/path_login.go [hunk 4]
```diff
@@ -1884,6 +1905,43 @@ func getMetadataValue(fromAuth *logical.Auth, forKey string) (string, error) {
...
 
+func awsRegionFromHeader(authorizationHeader string) (string, error) {
+	// https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-auth-using-authorization-header.html
+	// The Authorization header takes the following form.
+	//  Authorization: AWS4-HMAC-SHA256
+	//	Credential=AKIAIOSFODNN7EXAMPLE/20230719/us-east-1/sts/aws4_request,
+	// 	SignedHeaders=content-length;content-type;host;x-amz-date,
+	//	Signature=fe5f80f77d5fa3beca038a248ff027d0445342fe2855ddc963176630326f1024
+	//
+	// The credential is in the form of "<your-access-key-id>/<date>/<aws-region>/<aws-service>/aws4_request"
```
Comment (austingebauer) [inline builtin/credential/aws/path_login.go:1916]: Thanks, this comment is helpful!

File: builtin/credential/aws/path_config_client.go (hunk 1 above)
Comment (maxcoulombe) [reply builtin/credential/aws/path_config_client.go:65]: Spit-balling here, but would a verb prefix like `use_sts_region_from_client` clarify this is a boolean option and we do not mean "give us the region where the client will be calling from"?

For example the K8s secrets engine uses `disable_local_ca_jwt` not `local_ca_jwt` & PKI uses `allow_role_ext_key_usage` not `role_ext_key_usage` so there is precedent to using something like this on the API.

File: changelog/21960.txt [hunk 5]
```diff
@@ -0,0 +1,3 @@
+```release-note:improvement
+aws/auth: Adds a new config field `sts_region_from_client` which allows for using dynamic regional sts endpoints based on Authorization header when using IAM-based authentication..
```
Comment (maxcoulombe) [inline changelog/21960.txt:2]:
```suggestion
aws/auth: Adds a new config field `sts_region_from_client` which allows for using dynamic regional sts endpoints based on Authorization header when using IAM-based authentication.
```

Double dots?

File: website/content/api-docs/auth/aws.mdx [hunk 6]
```diff
@@ -65,6 +65,10 @@ capabilities, the credentials are fetched automatically.
 - `sts_region` `(string: "")` - Region to override the default region for making
   AWS STS API calls. Should only be set if `sts_endpoint` is set. If so, should
   be set to the region in which the custom `sts_endpoint` resides.
+- `sts_region_from_client` `(boolean: false)` - If set, overrides both `sts_endpoint`
+  and `sts_region` to instead use the region specified in the client request headers for
+  IAM-based authentication . This can be useful when you have client requests coming from
+  different regions and want flexibility in which regional STS API is used.
```
Comment (maxcoulombe) [inline website/content/api-docs/auth/aws.mdx:71]: Is it more typical to override other parameters or to enforce mutual exclusivity on the API?

I think for this particular use case going with an override sounds intuitive, but I'm wondering if Vault has an established preference that users may expect.

Comment (raymonstah) [reply website/content/api-docs/auth/aws.mdx:71]: That's a good callout. From what I can tell, overriding other parameters to change behavior doesn't seem to be a new concept in Vault. And searching for "override" in the codebase yields a more than a few results. It does seem to be a natural evolution of the API :)

File: builtin/credential/aws/path_config_client.go (hunk 1 above)
Comment (raymonstah) [reply builtin/credential/aws/path_config_client.go:65]: I have no objections to renaming it to `use_sts_region_from_client`, but I also want to call out that there exists other boolean fields that do not start with a verb. For example, the [LDAP Auth method](https://developer.hashicorp.com/vault/api-docs/auth/ldap#case_sensitive_names) has a few.

Comment (raymonstah) [reply builtin/credential/aws/path_config_client.go:65]: @hashicorp/docs & @hashicorp/vault-education-approvers, opinions appreciated <3

Comment (raymonstah) [reply builtin/credential/aws/path_config_client.go]: After some team feedback we've decided to name it to `use_sts_region_from_client`

File: website/content/api-docs/auth/aws.mdx (hunk 6 above)
Comment (maxcoulombe) [reply website/content/api-docs/auth/aws.mdx:71]: Gotcha, thanks for checking!

**Review by maxcoulombe — APPROVED** [review]
Thanks for the parameter rename. This is a really elegant solution, nice work.

**Pull request merged**